src/Renci.SshNet/Security/Cryptography/Ciphers/AesCipher.CtrImpl.cs

```diff
@@ -0,0 +1,230 @@
+
+            private byte[] CTREncryptDecrypt(byte[] data, int offset, int length)
+            {
+                var count = length / BlockSize;
+                if (length % BlockSize != 0)
+                {
+                    count++;
+                }
+
+                var buffer = new byte[count * BlockSize];
+                CTRCreateCounterArray(buffer);
+                _ = _encryptor.TransformBlock(buffer, 0, buffer.Length, buffer, 0);
+                ArrayXOR(buffer, data, offset, length);
+
+                // adjust output for non-blocksized lengths
+                if (buffer.Length > length)
+                {
+                    Array.Resize(ref buffer, length);
+                }
+
+                return buffer;
+            }
+
+#if NETSTANDARD2_1_OR_GREATER || NET6_0_OR_GREATER
+
+            // creates the Counter array filled with incrementing copies of IV
+            private void CTRCreateCounterArray(byte[] buffer)
+            {
+                var counter = MemoryMarshal.Cast<byte, uint>(buffer.AsSpan());
+
+                // fill array with IV, increment by 1 for each copy
+                var len = counter.Length;
+                for (var i = 0; i < len; i += 4)
+                {
+                    counter[i] = _packedIV[0];
+                    counter[i + 1] = _packedIV[1];
+                    counter[i + 2] = _packedIV[2];
+                    counter[i + 3] = _packedIV[3];
+
+                    // increment IV (little endian)
+                    if (_packedIV[3] < 0xFF000000u)
+                    {
+                        _packedIV[3] += 0x01000000u;
+                    }
+                    else
+                    {
+                        uint j = 3;
+                        do
+                        {
+                            _packedIV[j] = SwapEndianness(SwapEndianness(_packedIV[j]) + 1);
+                        }
+                        while (_packedIV[j] == 0 && --j >= 0);
+                    }
+                }
+            }
+
+            // XOR 2 arrays using Vector<byte>
+            private static void ArrayXOR(byte[] buffer, byte[] data, int offset, int length)
+            {
+                var vectorSize = Vector<byte>.Count;
+                for (var loopOffset = 0; length > 0; length -= vectorSize)
+                {
+                    if (length >= vectorSize)
+                    {
+                        var v = new Vector<byte>(buffer, loopOffset) ^ new Vector<byte>(data, offset + loopOffset);
+                        v.CopyTo(buffer, loopOffset);
+                        loopOffset += vectorSize;
+                    }
+                    else
+                    {
+                        for (var i = 0; i < length; i++)
+                        {
+                            buffer[loopOffset] ^= data[offset + loopOffset];
+                            loopOffset++;
+                        }
+                    }
+                }
+            }
```

Comment:
This can be:

```suggestion
            private static void ArrayXOR(byte[] buffer, byte[] data, int offset, int length)
            {
                int i = 0;
            
#if NETSTANDARD2_1_OR_GREATER || NET
                int oneVectorFromEnd = length - Vector<byte>.Count;
                
                for (; i <= oneVectorFromEnd; i += Vector<byte>.Count)
                {
                    var v = new Vector<byte>(buffer, i) ^ new Vector<byte>(data, offset + i);
                    v.CopyTo(buffer, i);
                }
#endif
                
                for (; i < length; i++)
                {
                    buffer[i] ^= data[offset + i];
                }
            }
```

Which would work on all platforms (the .NET Framework version looks like a deoptimization at first glance). This is not blocking feedback.

Reply:
Current ArrayXOR under Net Framework:
| Method      | Mean     | Error    | StdDev   | Gen0    | Allocated |
|------------ |---------:|---------:|---------:|--------:|----------:|
| Encrypt_CTR | 19.97 us | 0.398 us | 0.595 us | 11.8103 |  96.84 KB |
| Decrypt_CTR | 18.88 us | 0.378 us | 0.388 us | 11.8103 |  96.84 KB |

Your code under Net Framework:
| Method      | Mean     | Error    | StdDev   | Gen0   | Allocated |
|------------ |---------:|---------:|---------:|-------:|----------:|
| Encrypt_CTR | 29.97 us | 0.553 us | 0.517 us | 3.9978 |   32.8 KB |
| Decrypt_CTR | 29.73 us | 0.569 us | 0.585 us | 3.9978 |   32.8 KB |

I did mention in some other comment that I traded memory usage for performance for the NetFramework code. I think that's OK as it's a dying platform anyway.

I do like your Net7 code variant - the performance is the same but it's simpler. I'll commit it but will leave the NetFramework variant in place as it's faster.

Reply:
Ok, fair. I personally prefer simplicity and maintainability for older frameworks. We are also only considering 1 packet in isolation. When there are lots of packets on the wire, do we really want to be spending 3x the memory?

Reply:
These are short lived buffers, it's just a bit more work for the GC.

I ran a benchmark of encrypting 1024 x 32KB buffers (32 MB per iteration). In Net462, total active process memory hovered around 12 MB from start to finish. Under Net7 memory usage was actually higher at 22 MB. Both cases showed stable memory usage, looks like GC is doing a good job in both frameworks despite total allocated memory being 3x larger in Net462.